### PR TITLE
Feature-request #672 - database name prefixes + custom name

### LINKED ISSUE #672: [feature request] database name prefixes + custom name
I would like to choose the names for my databases.

I have now 7 databases called mostly like

user1
user2
...
user7

Is difficult to know which one is doing what
thx.

### PR DESCRIPTION
**Pull request by pquerner**
Add mysql prefix option to use customername + database description field as final databasename

Fixes #672 

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# How Has This Been Tested?

Locally. If the newely option value "DBNAME" is used, new functionality is triggered and executed.

# Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

I've only added english and german localization, as I do not speak the other languages.

### CODE REVIEW

File: lib/Froxlor/Api/Commands/Mysqls.php [hunk 1]
```diff
@@ -49,13 +49,23 @@ public function add()
...
 
 			// validation
 			$password = \Froxlor\Validate\Validate::validate($password, 'password', '', '', array(), true);
 			$password = \Froxlor\System\Crypt::validatePassword($password, true);
-			$databasedescription = \Froxlor\Validate\Validate::validate(trim($databasedescription), 'description', '', '', array(), true);
+			if(strtoupper(Settings::Get('customer.mysqlprefix')) == 'DBNAME') {
+			    //If DBNAME is used for prefix, we add customername + id + fieldescription to the validator.
+                //Also we pass in a regex, so we prevent invalid dbnames for mysql.
+                //read more @ https://dev.mysql.com/doc/refman/8.0/en/identifiers.html
+                $databasedescription = \Froxlor\Validate\Validate::validate($GLOBALS['userinfo']['loginname'] . $GLOBALS['userinfo']['customerid'] . '_' . trim($databasedescription), 'description', '/^[0-9a-zA-Z$_]{1,64}$/', '', array(), true);
```
Comment (d00p) [inline lib/Froxlor/Api/Commands/Mysqls.php]: $GLOBALS ???? Really??? Please use `$this->getUserDetail('FIELD')`

Also, why additionally put the customer-id in there? Loginame is unique. The database name cannot be any length, there are restrictions

File: lib/Froxlor/Database/DbManager.php [hunk 2]
```diff
@@ -79,14 +79,17 @@ public function createDatabase($loginname = null, $password = null, $last_accnum
...
+            $allsqlusers = $this->getManager()->getAllSqlUsers();
+            // generate random username
+            $username = $loginname . '-' . substr(md5(uniqid(microtime(), 1)), 20, 3);
+            // check whether it exists on the DBMS
+            while (in_array($username, $allsqlusers)) {
+                $username = $loginname . '-' . substr(md5(uniqid(microtime(), 1)), 20, 3);
+            }
+        }
+        if (strtoupper(Settings::Get('customer.mysqlprefix')) == 'DBNAME') {
+            $username = $GLOBALS['userinfo']['loginname'] . $GLOBALS['userinfo']['customerid'] . '_' . $GLOBALS['_POST']['description'];
```
Comment (d00p) [inline lib/Froxlor/Database/DbManager.php]: `$GLOBALS['_POST']['description']` ??? You even adjusted the parameter above....it's just `$databasedescription`
Please never ever use $GLOBALS, we don't this here...

File: lib/Froxlor/Api/Commands/Mysqls.php (hunk 1 above)
Comment (pquerner) [reply lib/Froxlor/Api/Commands/Mysqls.php]: The 64 char length is done via the Regexp.
I'll take care of not using GLOBALS - I did not look through the other sourcecode beforehand and chose to just use GLOBALS to get the PR out and then probably get shown the correct way.
I will remove the customer-id.

**Comment by pquerner** [comment]
I've added a new commit to this
https://github.com/Froxlor/Froxlor/pull/680/commits/050260705a9288c80181fc2948cf9515848d46ec

**Comment by pquerner** [comment]
Is there anything else I should change or is missing @d00p  ?

File: lib/Froxlor/Api/Commands/Mysqls.php [hunk 3]
```diff
@@ -49,13 +49,28 @@ public function add()
...
+
+            // get needed customer info to reduce the mysql-usage-counter by one
+            $customer = $this->getCustomerData('mysqls');
+			$newdbname = '';
+			if(strtoupper(Settings::Get('customer.mysqlprefix')) == 'DBNAME') {
+			    //If DBNAME is used for prefix, we add customername + id + fieldescription to the validator.
+                //Also we pass in a regex, so we prevent invalid dbnames for mysql.
+                //read more @ https://dev.mysql.com/doc/refman/8.0/en/identifiers.html
+                $newdbname = ($this->isAdmin() ? $customer['loginname'] : $this->getUserDetail('loginname')) . '_' . trim($databasedescription);
+                $databasedescription = \Froxlor\Validate\Validate::validate($newdbname, 'description', '/^[0-9a-zA-Z$_]{1,64}$/', '', array(), true);
```
Comment (d00p) [inline lib/Froxlor/Api/Commands/Mysqls.php]: shouldn't that be:
```php
$newdbname = \Froxlor\Validate\Validate::validate($newdbname, 'description', '/^[0-9a-zA-Z$_]{1,64}$/', '', array(), true);
```
Which automatically makes the variable `$originaldatabasedescription` superfluous

File: lng/english.lng.php [hunk 4]
```diff
@@ -329,7 +329,7 @@
 $lng['serversettings']['accountprefix']['title'] = 'Customer prefix';
 $lng['serversettings']['accountprefix']['description'] = 'Which prefix should customer accounts have?';
 $lng['serversettings']['mysqlprefix']['title'] = 'SQL Prefix';
-$lng['serversettings']['mysqlprefix']['description'] = 'Which prefix should MySQL accounts have?</br>Use "RANDOM" as value to get a 3-digit random prefix';
+$lng['serversettings']['mysqlprefix']['description'] = 'Which prefix should MySQL accounts have?</br>Use "RANDOM" as value to get a 3-digit random prefix.</br>Use "DBNAME" as value to use mysql description field as db name (together with customer name and id prefix).';
```
Comment (d00p) [inline lng/english.lng.php]: customer id reference should be removed as it is not used anymore

File: lng/german.lng.php [hunk 5]
```diff
@@ -325,7 +325,7 @@
 $lng['serversettings']['accountprefix']['title'] = 'Kundenpräfix';
 $lng['serversettings']['accountprefix']['description'] = 'Welchen Präfix sollen die Kundenaccounts haben?';
 $lng['serversettings']['mysqlprefix']['title'] = 'MySQL-Präfix';
-$lng['serversettings']['mysqlprefix']['description'] = 'Welchen Präfix sollen die MySQL-Benutzerkonten haben?</br>Mit "RANDOM" als Wert wird ein 3-stelliger Zufallswert als Präfix verwendet.';
+$lng['serversettings']['mysqlprefix']['description'] = 'Welchen Präfix sollen die MySQL-Benutzerkonten haben?</br>Mit "RANDOM" als Wert wird ein 3-stelliger Zufallswert als Präfix verwendet.</br>Mit "DBNAME" als Wert wird der Datenbankbeschreibungstext zusammen mit dem Kundennamen und KundenID als Präfix genutzt.';
```
Comment (d00p) [inline lng/german.lng.php]:
```diff
-- der Datenbankbeschreibungstext
++ die Datenbankbeschreibung
```
und auch "Kunden-ID" Referenze aus dem string entfernen

**Comment by makuser** [comment]
Woran hängt es hier?

File: lib/Froxlor/Api/Commands/Mysqls.php (hunk 3 above)
Comment (pquerner) [reply lib/Froxlor/Api/Commands/Mysqls.php]: Not sure anymore why I made this, I've removed it.

**Comment by d00p** [comment]
Closed in favor of #956

**Pull request closed without merging**